### ISSUE TITLE
rucio.cfg config options for OIDC. New Feature: automatic token refresh on the client side. 

### ISSUE THREAD

**Issue by paulmillar** [rucio/rucio#3572]
Motivation
----------

OIDC support in the Rucio client currently requires the user either to supply the OP's authentication password with each Rucio invocation, or to record the password (unencrypted) in the `rucio.cfg` file.  Neither solution is particularly appealing.

The OIDC protocol supports refresh tokens.  Storing a refresh token would allow the Rucio client to request a fresh access token when needed, without requiring the user to re-authenticate (typically with username + password).

The rucio client could implement support for obtaining and storing a refresh token, and for obtaining an access token when necessary; however, the oidc-agent project (see https://github.com/indigo-dc/oidc-agent) has already done this.  It has also done this in a way that is convenient for users.  Additionally, oidc-agent provides a python client (see https://github.com/indigo-dc/liboidc-agent-py) that should make it easy to include support for refresh tokens.

Modification
------------

Add the option for the rucio client to use a specific oidc-agent account when authenticating with Rucio.  Include the oidc-agent python client and update the REST dispatch to obtain an up-to-date access token from oidc-agent.

**Comment by bari12** [comment]
@gumond can you please have a look. I think this workflow already partially exists insofar that Rucio is able to use a refresh token to renew the access of the user. I am not sure though if you can specify a refresh token in the first place yourself though.

**Comment by gumond** [comment]
Hello Paul, 

Thank you for your proposal.


Security wise: 

The refresh token is very sensitive information, that needs to be stored very securely (just as the password), this is one of the reasons, why Rucio server does not return to the Rucio client the refresh token despite offline access scope was requested in the CLI at the time authentication was requested by the user. Rucio server only stores this token for a server side use in the Rucio DB.  From the security perspective, I do see at least the same problem in saving refresh token in the rucio.cfg file as when saving there the password. We might consider it in cases one could make sure that the refresh tokens have a limited lifetime and are usable only once e.g., but there is still a  question of this authentication principle/flow being correct in principle. The principle seems always correct to me when password is used, but we can ask some OIDC/OAuth2 experts to be sure. 

Convenience:

I do understand this could be a convenient option (and it would not be hard to implement), but it sounds a bit like bypassing the authentication code flow and making out of the OIDC/OAuth2.0 a simple userpass case. If we come to a general agreement, this is what we want/need, we can surely implement it, but at this point, I am unsure this is a correct use case. 


What about the following proposals:

a) if 1h token on the client side is not enough, we could try to have longer token lifetimes configured on the IAM side

b) if you are worried about operations performed on the server side on behalf of the user, using OIDC authentication, this is covered already - see "Regarding token refresh below"

c) in case Rucio user wishes to pre-configure rucio.cfg file to use auth_type = oidc, the Rucio client could automatically initiate login through 'polling'/'fetch code' strategies and prompt the user on the CLI. In such cases, user would never type the password in the CLI (Rucio client would never see it). In this way, the authentication stays secure with no compromises. Despite not invoked automatically, you can now achieve this by simply typing e.g.: 
_rucio -a=<account> -s OIDC  -v whoami_ 
anytime your current authentication expires. 

d) I imagine Rucio client tries to authenticate you automatically once your token expires (with your IdP username & password), please let me know if this is not the case (assuming oidc-auto is True and all information is filled in the rucio.cfg file).


Regarding token refresh in general: 

Currently a user can authenticate via the CLI requiring 'offline_access' scope  and optionally --oidc-refresh-lifetime. In this way, Rucio will automatically continue refreshing the token for the any operations user needs Rucio to perform even after the user is no longer having authenticated session on the client side (assuming oauth_manager daemon is running on the server side - which is the case on the Rucio DOMA instance now).

I might have misunderstood your proposed use-case. Please let me know if anything is unclear. We might want to think of discussing this proposal with IdP (IAM) experts as well.

**Comment by paulmillar** [comment]
Hi @gumond , thanks for the quick reply.

To be honest, I was surprised and alarmed when, reading through the Rucio docs, it seemed that I must supply Rucio (either the client or the server) my OIDC credentials: my IAM username and password.  The OIDC client (whether that is the Rucio client or the Rucio server) really, really shouldn't need this information.  It's almost the defining reason why OAuth and OAuth2 (from which OIDC is derived) exists: that a third-party is authorised without requiring the resource-owner's credentials.

(if I can authn with OIDC and without telling either the Rucio server or client my OIDC password then please let me know how!)

The refresh token I mentioned is actually an implementation detail.  What I would like is that the Rucio client can use OIDC authentication (without requiring me to enter my OIDC password) until I navigate to the IAM instance and revoke the permission where I granted the OIDC-client access to information about myself.  [This page](https://rclone.org/webdav/#openid-connect) describes how to do this with dCache and a WebDAV client called rclone, using oidc-agent.  If you don't know oidc-agent, I *strongly* recommend you investigate what it does and how it works; if only to appreciate an alternative way of achieving the same goal.

My concrete proposal is that the Rucio client uses oidc-agent to obtain an OIDC access token when authenticating with the Rucio server.  The Rucio client then uses this access token when making REST API calls.  However, this is just an example of how this could work.

I like the idea of involving an OIDC expert in this discussion.  If I may recommend, we should involve @andreaceccanti who wrote the IAM and is quite knowledgable on OIDC.  It would also be help if there was something that documented exactly how Rucio works with OIDC (a power-point presentation would be fine) within which there's a description of where the different oidc-* options have an influence.

Cheers,
Paul.

**Comment by gumond** [comment]
Hi Paul, 

again, there are 3 ways to authenticate via the CLI. 2 of them do not require you to pass any sensitive information to Rucio client at all. These 2 will use your web browser in the process. When implementing these 3 approached in 2 of them we did get inspired by the oidc-agent you write about. 

(if I can authn with OIDC and without telling either the Rucio server or client my OIDC password then please let me know how!)

Yes you certainly can, please, follow examples here: 

Login via user’s browser + fetch code:
rucio -a=<rucio_account_name> -S=OIDC -v whoami
Login via user’s browser + polling Rucio auth server:
rucio -a=<rucio_account_name> -S=OIDC --oidc-polling -v whoami

https://rucio.readthedocs.io/en/latest/cli_examples.html

Please let me know if you do not succeed, I will be happy to help !

------------------------------------------------------------------------

I am sorry to read, you were alarmed. We included the option to trust Rucio Client with the password after it was discussed on several meetings. We can, of course, discuss removing it, if there are strong reasons to. That said, users are free not to trust Rucio Client with the password and use the other 2 CLI authentication strategies.  As you are strongly for pure OIDC/OAuth2, I really think trusting Rucio Client (or oidc-agent) with refresh token is a very similar problem to the one of trusting it with a password. The refresh token is basically your identity and it is not just an implementation detail AFAICS. 

"revoke the permission where I granted the OIDC-client access to information about myself"

In the 2 CLI examples I mention above, you will see this page in your browser (unless you used the --oidc-auto already in which case the Rucio Client will click on this page for you). It should be sufficient to follow the instructions on the command line. Let me know if this works on you. We can also chat on Slack if you have troubles with it.  

Your proposal, use oidc-agent, was something we considered before implementing OIDC in our Rucio Client. There is always room for reimplementation or rethinking the current strategy, but I strongly believe that the current implementation on the Rucio Client side does the right thing and I do not see reasons for using oidc-agent, but again if you give us strong reasons for a change we will be happy to, but let's try first to understand what is actually now in place:
 
 -> Rucio Client asks the Rucio auth server for the authorization request url (specific to the user & Rucio auth server IAM client)
-> user authorises with this request url through browser
(in case you do trust Rucio Client with your password, it proceeds automatically and it also clicks "allow" on the "revoke permissions page" you mentioned - again you are free not to use this functionality !) 
-> user sees a fetch code in the browser and copy-pastes it to the Rucio Client through the instructions on the CLI (again there are 2 other options in which the Rucio Client polls the Rucio server for the token after authentication succeeded with the IdP or the automatic option in which Rucio Client gets directly the token and no browser interaction is initiated)

Feel free to involve Andrea, I am in contact with him on Slack as well (which might be easier to resolve these exchanges/misunderstandings), but before we do include more people, I would like to understand, what in particular you see missing and why do you so strongly want to see oidc-agent being used on the Rucio Client side. I think we covered most if not all the use cases with minimal code on the Rucio Client side.   

The presentation of how Rucio works with OIDC was made, but of course, I can make more slides if you wish in the future (if I am still at CERN). The documentation is available in Rucio docs: 

https://rucio.readthedocs.io/en/latest/cli_examples.html#open-id-connect-authentication-examples
https://rucio.readthedocs.io/en/latest/installing_server.html#server-configuration-for-open-id-connect-authn-z

I am still missing what do you miss in terms of functionality and what you see incorrect for which we are so strongly encouraged to change client side implementation (and possibly rethink the server implementation as well).

------------------------------------------------------------------------

As I see it, this thread started as a request for "convenience" feature. And I find it a good idea to think about ! I might have yet another proposal in which a user after first successful authentication via one of the 3 CLI methods could be automatically re-authenticated (Rucio Client polling auth server for token that was refreshed on the server side using the already existing refresh mechanism/daemon and minimal additions to the Rucio Client).

**Comment by paulmillar** [comment]
Hi Jaroslav,

Thanks for the information and help.  I was able to get both the "Login via user’s browser + fetch code"
and the "Login via user’s browser + polling Rucio auth server" methods to work for me.

I noticed that these methods result in an access token being stored (for me) as `/tmp/paul/.rucio_pmillar/auth_token_pmillar`.

As an aside, with the following command, I was able to create a workable file from oidc-agent:

```
oidc-token --aud=rucio XDC-IAM | tr -d '\n' > /tmp/paul/.rucio_pmillar/auth_token_pmillar
```

(The `tr -d '\n'` is needed as rucio client takes the file's contents, as-is, and inserts it into the HTTP header.  This only works if the file has no newline characters.)

I found another oddity when placing options in the `rucio.cfg` file.

With the following configuration, the rucio client works, but requires that I specify `-S oidc` on the command-line:
```
[client]
rucio_host = https://rucio-doma.cern.ch
auth_host = https://rucio-doma-auth.cern.ch
ca_cert = /home/paul/.globus/IGTF/CERN-bundle.crt
account = pmillar
```

When I try to configure the authentication type via the `auth_type = oidc` statement (the value `OIDC` is not accepted) suddenly all `oidc_` properties (`oidc_auto`, `oidc_scope`, ...) must be specified.  Like this, I was unable to get the two OIDC options (that don't require recording the OIDC credential) to work.

Yes, you're quite right about the refresh token representing my identity.  The same is true for my gpg key (as stored by gpg-agent), my ssh key (as stored by ssh-agent) and my grid/voms proxy file.  These agents/files provide a trade-off between how well my identity has been asserted against convenience.   In all cases, the long-lived credential is stored encrypted and an "unlock" version is stored  for  the duration of a login session (or shorter), or is only valid for a day.

So, in summary, it might be useful if the rucio client were to take an access token from oidc-agent, but it is certainly not urgent.

**Comment by andreaceccanti** [comment]
Hello @gumond, @paulmillar ,

I completely support Paul's arguments: having the user password stored in clear in the RUCIO config file is not ideal and goes against the basic principles behind OAuth/OIDC. OAuth provides refresh tokens (RTs) to support offline access scenarios. RTs are client specific, can be revoked, can be time-limited, provide access only to the scopes authorized for the client, i.e. there are many good reasons to use RTs :)

I would also recommend embedding oidc-agent in the RUCIO cli, or leave this responsibility out of the RUCIO cli. This would mean that before useing the RUCIO cli, the user has to obtain a token in some way. The token could then be exposed to the RUCIO cli via an environment variable, as we do today for X.509 proxy certificates. Note that a proposal for the name and meaning of this environment variable has been discussed in the context of the WLCG AuthZ WG, and presented at the latest GDB (slides are [here][gdb-06052020]). It would be great to have support for this in the RUCIO cli!

[gdb-06052020]: https://indico.cern.ch/event/813747/contributions/3841759/attachments/2032485/3402000/20200506_GDB_AuthZ_Update.pptx.pdf

**Comment by gumond** [comment]
Hello @andreaceccanti , 

> I completely support Paul's arguments: having the user password stored in clear in the RUCIO config file is not ideal and goes against the basic principles behind OAuth/OIDC.

This was also my argument and as I explained, clearly I hope, there are several ways to login with Rucio Client. User is not, by any means, forced to store his password anywhere in the Rucio Client. Please contact me if you wish to try what has actually been implemented and how to use it if the documentation I point to is not clear enough. 

> there are many good reasons to use RTs :)

Of course, I understood this and at any point, I did not refuse to consider the convenience feature request of Paul. On the other hand, RTs stay a very sensitive information and the way we would use them is something to do carefully. In addition, the RT is stored on the server side which also does the token refresh automatically. The only step to do in case this refresh is to become automatic on the client side is to implement that the client asks for a new token before the current one expires, but that is really convenience use-case that we can discuss in this thread. That said, of course at any point, you can use any valid token (with the Rucio server requested audience and scopes) through the REST API.

> I would also recommend embedding oidc-agent in the RUCIO cli:

It is nice you do recommend it, but I am sorry, I do not read on what grounds. The Rucio Client currently manages the same use-cases and I do not see what you are urgently missing or see unfit the current use-cases. Please before we suggest changes, you may want to understand better what was done. 

> before using the RUCIO cli, the user has to obtain a token in some way.:

This is also possible. If you read the discussion above, you may see this is also what Paul has figured out works eventually (I believe). 

> The token could then be exposed to the RUCIO cli via an environment variable: 

Really not difficult to put in place. 

Please, this conversations is getting a bit out of scope of this request and I would propose that we get together on Slack or in a video chat where I explain you what currently works and you give me feedback on missing pieces or urgently necessary changes. For the moment, I do not see any strong arguments for oidc-agent as all the use-cases you describe are mostly covered and first one should test those before trying out to propose changes or pushing for tools just because you fancy them. 

Cheers, 
Jaroslav

**Comment by gumond** [comment]
Hi @andreaceccanti, @paulmillar, 

to summarise:

feature requests: 

* automatic re-logging of Rucio Client when offline_access has been granted to Rucio
* support for access to a token via env variables as stated in the slides from last week https://indico.cern.ch/event/813747/contributions/3841759/attachments/2032485/3402000/20200506_GDB_AuthZ_Update.pptx.pdf

requested fixes: 

* problems with pre-configuring Rucio Client OIDC authentication from the rucio.cfg file. Current  optional vs obligatory relationship of the parameters do not allow to use this configuration mechanism for using all authentication methods. Once this is fixed all 3 CLI authentication methods should become available also from the rucio.cfg config file. Note: there is also the a 4th option of providing a valid token in a Rucio Client & user specific file. 

AFAIU,  this is what you need us to improve and it is definitely something we can do without replacing Rucio Client OIDC capabilities with oidc-agent which does very much the same thing as we provide at the moment. 

Thank you for testing things out and for your feedback !

**Comment by andreaceccanti** [comment]
HI @gumond,

> Of course, I understood this and at any point, I did not refuse to consider the convenience 
> feature request of Paul. On the other hand, RTs stay a very sensitive information and the way we would use them is something to do carefully. 

The password credential is much more sensitive, and it's currently being stored in plain text.
Even storing the RT in plain text would be an improvement over what you have now.

> The Rucio Client currently manages the same use-cases and I do not see what you are urgently missing or see unfit the current use-cases. 

Well, it does not support refresh tokens, so it does not support the same use cases.
If using oidc-agent is not convenient for you that's fine! I was suggesting it since I find it pretty convenient.

Thanks!

PS: you can quote previous comments by using the `>` prefix, as in email!

**Comment by gumond** [comment]
Hi @andreaceccanti , 

> The password credential is much more sensitive, and it's currently being stored in plain text.
Even storing the RT in plain text would be an improvement over what you have now.

I am sorry, but why do I have to write x-times here that nobody forces the user to type a password and store it in rucio.cfg ? There are other ways available as Paul found out as well. Improvements and fixes will come as I summarised. 

The argument for implementing this option (saving a password on the client side) was made on several DOMA meetings. It was pointed out that some current workflows will need this use-case and it was decided to implement it. If we find out other ways to make these workflows work, I am fine removing this option as I do understand it is in principle against OIDC/OAuth2 and it should not be there. So this is something to follow up on. 

> Well, it does not support refresh tokens, so it does not support the same use cases.

Again, this is a convenience feature that we can add. 

> PS: you can quote previous comments by using the > prefix, as in email!
Thanks ;) 

I hope I summarised the status of the features requested in addition + fixes to come. 
If there is more you would want from us, please let us know !

**Comment by paulmillar** [comment]
Hi guys,

Sorry for the confusion.  It *is* possible to run the rucio client with OIDC authentication and without specifying the password.  My confusion was partly due to a problem with how the rucio client is configured, for which I've opened a fresh ticket (see #3581).

Once authenticated, the client stores an access token (AT) and uses this to authenticate with the rucio server.  This AT has whatever lifetime the OP gives it.  Once it has expired, the rucio client prompts the user to re-authenticate.  This is annoying because it requires browser interaction.

Therefore, the advantage of using a refresh token (RT) is that this is avoided: when the AT expires, the RT is used to fetch a new AT.  Using oidc-agent would mean the rucio client does not have to handle storing long-lived credentials.  I have created the rucio client's access token from oidc-agent and verified this works with the rucio server.  Therefore, integrating oidc-agent should be relatively easy.

I'd favour direct integration (where the rucio client requests the AT from oidc-agent) over an environment variable as this avoids the AT refresh problem, where the environment variable needs to be updated when the AT expires.

As an aside, I think having oidc-user and oidc-password as options is wrong.  It is far too easy for people to "do the wrong thing" and leak credentials (copy-n-paste command-lines; hard-coding values in scripts, etc).  However, I'm happy if I can run the rucio client without specifying my OIDC credential.

**Comment by andreaceccanti** [comment]
>I'd favour direct integration (where the rucio client requests the AT from oidc-agent) over an environment variable as this avoids the AT refresh problem, where the environment variable needs to be updated when the AT expires.

Just to mention that the AuthZ WG proposal describes the `BEARER_TOKEN_FILE` env variable, which points to a file where an access token lives for the current user. Using that approach would avoid env variable value update problems.

**Comment by gumond** [comment]
Hi Paul,

Regarding the re-authentication via refresh tokens. Sure, we can figure out a solution ! 
The problem with RTs can be that there is a Rucio client-server relationship that should be preserved and the management of RTs issued through the server side should stay on the server side by principle. Having something for the client is surely possible and might not be difficult. Let us think about it a bit. 

> As an aside, I think having oidc-user and oidc-password as options is wrong. It is far too easy for people to "do the wrong thing" and leak credentials (copy-n-paste command-lines; hard-coding values in scripts, etc).

This was extensively discussed many month ago and can be rediscussed. We could make it less accessible or remove it of course once we find solutions for all our use-cases which currently need this option. 

Cheers, 
Jaroslav

**Comment by gumond** [comment]
Hi Andrea again, 

BEARER_TOKEN_FILE env variable is something  we can certainly implement as well.  
We can think of reimplementation of the current Rucio OIDC altogether following your suggestions as well, but I would focus on first trying to address things you see as showstoppers in what we currently have (e.g. the rucio.cfg file oidc option specification fix). 

Note: I recognise that the following message was not correct to you and I apologise for that, there was no intention in offending you: 

> I am sorry, but why do I have to write x-times here that nobody forces the user to type a password and store it in rucio.cfg ? There are other ways available as Paul found out as well. Improvements and fixes will come as I summarised.

That said, I proposed discussion over video chat or Slack where these big misunderstandings could have been avoided and we could have used GitHub only for the relevant information to push the project forward.

**Comment by andreaceccanti** [comment]
> Note: I recognise that the following message was not correct to you and I apologise for that, there was no intention in offending you:

NP Jaroslav, i think we are on the same page!

**Comment by gumond** [comment]
Yes, I think now we are.

The problem Paul faces comes from the fact, that I thought the use-case of fully automatic authentication, pre-configurable from the rucio.cfg, is feasible only through user's password (the method agreed on the DOMA meetings). That is why, once you try to configure OIDC through this file it currently forces a user to enter all relevant parameters, password included. In addition to this option, I though a user would mostly use the (3) CLI authentication options where 2 are not automatic, but require user's action in any case. 

I can understand that in principle and for convenience, the defaults should be presettable from the rucio.cfg for all the methods and that we could put in place some other ways to automate the re-authentication process to avoid user being prompted after token expiration. 

I am happy Paul, found a work-around via the 3 other CLI options which work and even found a 4th option as a work-around in writing his token to the Rucio Client token file directly. 

I will try to push these new features to the next Rucio release.
